From efb1526fdf5efa1d259109451163abbeabe79a31 Mon Sep 17 00:00:00 2001
From: Michael Clark <mjc@sifive.com>
Date: Thu, 1 Feb 2018 15:01:26 +1300
Subject: [PATCH 2/2] RISC-V - Implement priv ISA v1.10 trap and sret/mret
 xPIE/xIE behavior

1.9.1 : When a trap is taken from privilege mode y into privilege mode
x, xPIE is set to the value of y IE; x IE is set to 0; and xPP is set
to y.

1.10: When a trap is taken from privilege mode y into privilege mode
x, xPIE is set to the value of x IE; x IE is set to 0; and xPP is set
to y.
---
 target/riscv/helper.c    |  6 ++++--
 target/riscv/op_helper.c | 12 ++++++++----
 2 files changed, 12 insertions(+), 6 deletions(-)

diff --git a/target/riscv/helper.c b/target/riscv/helper.c
index 9e4c5efed4..a55e204b33 100644
--- a/target/riscv/helper.c
+++ b/target/riscv/helper.c
@@ -431,7 +431,8 @@ void riscv_cpu_do_interrupt(CPUState *cs)
         }
 
         target_ulong s = env->mstatus;
-        s = set_field(s, MSTATUS_SPIE, get_field(s, MSTATUS_UIE << env->priv));
+        s = set_field(s, MSTATUS_SPIE, env->priv_ver >= PRIV_VERSION_1_10_0 ?
+            get_field(s, MSTATUS_SIE) : get_field(s, MSTATUS_UIE << env->priv));
         s = set_field(s, MSTATUS_SPP, env->priv);
         s = set_field(s, MSTATUS_SIE, 0);
         csr_write_helper(env, s, CSR_MSTATUS);
@@ -451,7 +452,8 @@ void riscv_cpu_do_interrupt(CPUState *cs)
         }
 
         target_ulong s = env->mstatus;
-        s = set_field(s, MSTATUS_MPIE, get_field(s, MSTATUS_UIE << env->priv));
+        s = set_field(s, MSTATUS_MPIE, env->priv_ver >= PRIV_VERSION_1_10_0 ?
+            get_field(s, MSTATUS_MIE) : get_field(s, MSTATUS_UIE << env->priv));
         s = set_field(s, MSTATUS_MPP, env->priv);
         s = set_field(s, MSTATUS_MIE, 0);
         csr_write_helper(env, s, CSR_MSTATUS);
diff --git a/target/riscv/op_helper.c b/target/riscv/op_helper.c
index 00305b5c70..5d986dcc62 100644
--- a/target/riscv/op_helper.c
+++ b/target/riscv/op_helper.c
@@ -612,8 +612,10 @@ target_ulong helper_sret(CPURISCVState *env, target_ulong cpu_pc_deb)
 
     target_ulong mstatus = env->mstatus;
     target_ulong prev_priv = get_field(mstatus, MSTATUS_SPP);
-    mstatus = set_field(mstatus, MSTATUS_UIE << prev_priv,
-                        get_field(mstatus, MSTATUS_SPIE));
+    mstatus = set_field(mstatus,
+        env->priv_ver >= PRIV_VERSION_1_10_0 ?
+        MSTATUS_SIE : MSTATUS_UIE << prev_priv,
+        get_field(mstatus, MSTATUS_SPIE));
     mstatus = set_field(mstatus, MSTATUS_SPIE, 0);
     mstatus = set_field(mstatus, MSTATUS_SPP, PRV_U);
     riscv_set_mode(env, prev_priv);
@@ -635,8 +637,10 @@ target_ulong helper_mret(CPURISCVState *env, target_ulong cpu_pc_deb)
 
     target_ulong mstatus = env->mstatus;
     target_ulong prev_priv = get_field(mstatus, MSTATUS_MPP);
-    mstatus = set_field(mstatus, MSTATUS_UIE << prev_priv,
-                        get_field(mstatus, MSTATUS_MPIE));
+    mstatus = set_field(mstatus,
+        env->priv_ver >= PRIV_VERSION_1_10_0 ?
+        MSTATUS_MIE : MSTATUS_UIE << prev_priv,
+        get_field(mstatus, MSTATUS_MPIE));
     mstatus = set_field(mstatus, MSTATUS_MPIE, 0);
     mstatus = set_field(mstatus, MSTATUS_MPP, PRV_U);
     riscv_set_mode(env, prev_priv);
-- 
2.15.1

